tables.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package winmd
+
+import (
+	"github.com/microsoft/go-winmd/flags"
+)
+
+// Record is an item in a metadata table.
+type Record = any
+
+// Index indexes a record of type T.
+type Index[T Record] int
+
+// CodedIndex indexes a record on any table.
+type CodedIndex struct {
+	Index[any]
+	Table int
+}
+
+// Slice indexes the range of records [Start,End) on the table T.
+type Slice[T Record] struct {
+	Start Index[T]
+	End   Index[T]
+}
+
+// BlobIndex is an index to a #Blob heap.
+type BlobIndex int
+
+// GuidIndex is an index to a #GUID heap.
+type GUIDIndex int
+
+// Assembly is defined in §II.22.2.
+type Assembly struct {
+	HashAlgID      flags.AssemblyHashAlgorithm
+	MajorVersion   uint16
+	MinorVersion   uint16
+	BuildNumber    uint16
+	RevisionNumber uint16
+	Flags          flags.AssemblyFlags
+	PublicKey      BlobIndex
+	Name           string
+	Culture        string
+}
+
+// assemblyOS is defined in §II.22.3.
+// This record should not be emitted into any PE file,
+// but needed temporarily to calculate sizes and offsets for subsequent tables.
+type assemblyOS struct {
+	OSPlatformID   uint32
+	OSMajorVersion uint32
+	OSMinorVersion uint32
+}
+
+// assemblyProcessor is defined in §II.22.4.
+// This record should not be emitted into any PE file,
+// but needed temporarily to calculate sizes and offsets for subsequent tables.
+type assemblyProcessor struct {
+	Processor uint32
+}
+
+// AssemblyRef is defined in §II.22.5.
+type AssemblyRef struct {
+	MajorVersion     uint16
+	MinorVersion     uint16
+	BuildNumber      uint16
+	RevisionNumber   uint16
+	Flags            flags.AssemblyFlags
+	PublicKeyOrToken BlobIndex
+	Name             string
+	Culture          string
+	HashValue        BlobIndex
+}
+
+// assemblyRefOS is defined in §II.22.6.
+// This record should not be emitted into any PE file,
+// but needed temporarily to calculate sizes and offsets for subsequent tables.
+type assemblyRefOS struct {
+	OSPlatformID   uint32
+	OSMajorVersion uint32
+	OSMinorVersion uint32
+	AssemblyRef    Index[AssemblyRef]
+}
+
+// assemblyRefProcessor is defined in §II.22.7.
+// This record should not be emitted into any PE file,
+// but needed temporarily to calculate sizes and offsets for subsequent tables.
+type assemblyRefProcessor struct {
+	Processor   uint32
+	AssemblyRef Index[AssemblyRef]
+}
+
+// ClassLayout is defined in §II.22.8.
+type ClassLayout struct {
+	PackingSize uint16
+	ClassSize   uint32
+	AssemblyRef Index[TypeDef]
+}
+
+// Constant is defined in §II.22.9.
+type Constant struct {
+	Type    flags.ElementType
+	Padding byte       // 1-byte padding zero
+	Parent  CodedIndex `winmd:"HasConstant"`
```

Comment:
In theory we could take care of the type check on the library side and provide type-safe usage with some sort of visitor, right?

```go
type HasConstantIndex = CodedIndex

func (i HasConstantIndex) Get(
	tables *BunchOfTables,
	field func(f *Field) error,
	param func(p *Param) error,
	property func(p *Property) error) error {

	x, err := tables.Get(i)
	if err != nil {
		return err
	}
	switch v := x.(type) {
	case *Field:
		return field(v)
	case *Param:
		return param(v)
	case *Property:
		return property(v)
	}
	return fmt.Errorf("unexpected type in table, not HasConstant: %v", x)
}
```

This way as a user you know you're checking the possible subtypes exhaustively by filling in every parameter. Although I don't think this is a particularly friendly API... especially for `HasCustomAttribute` with its 22 possibilities! 😄

(I suppose you could generalize this approach by writing `CodedIndex2[T1, T2 any]`, `CodedIndex3[T1, T2, T3 any]`, ..., but that would only help on the library writing side, not usage, as far as I can tell.)

It seems fairly normal to have to do type assertion switches in Go (https://go.dev/doc/faq#variant_types, https://github.com/golang/go/issues/19412), so I'm uncertain doing all this work in the library would be worth the trouble, though.

Reply:
I tried a lot of things before assuming that good-old type assertion is the right things in this case. As you mention, other options complicate the API and even client usage.